_published/0014-test-results-handling.md

```diff
@@ -0,0 +1,256 @@
+For this we will need to also store the messages that the tests produce.
+At the very least for failed tests.
+
+## Motivation
+
+This is bad for the user because it leads to slower response times when
+loading the project view. It's also bad for the API because it leads to
+unnecessary fetches, and pointless repeated computation.
+
+Having the tests' error messages is pretty much a necessity when trying
+to fix them.
+
+## Related GitHub issues
+- wharf-api [#11](https://github.com/iver-wharf/wharf-api/issues/11) [#17](https://github.com/iver-wharf/wharf-api/issues/17)
+- wharf-web [#13](https://github.com/iver-wharf/wharf-web/issues/13)
+
+## Explanation
+
+<details><summary>wharf-api</summary>
+
+The POST `/build/{buildid}/artifact` endpoint handles inserting artifacts.
+If there are TRX (XML) files, it also parses them to create an array of `TestResult` and
+one `TestResultSummary` per file.
+
+The summaries get inserted into the database table `test_result_summary`.
+The results get inserted into the database table `test_result`.
+
+![Database structure](https://github.com/iver-wharf/rfcs/blob/33e14dceb44755ce5e0c66b257bbffe2bb3568d7/_assets/wharf-db-graph.png)
+
+Pseudocode-like, without error handling
+```go
+// database_models.go
+// modified
+type Build struct {
+    BuildID     uint         `gorm:"primaryKey" json:"buildId"`
+    StatusID    BuildStatus  `gorm:"not null" json:"statusId"`
+    ProjectID   uint         `gorm:"not null;index:build_idx_project_id" json:"projectId"`
+    Project     *Project     `gorm:"foreignKey:ProjectID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+    ScheduledOn *time.Time   `gorm:"nullable;default:NULL" json:"scheduledOn" format:"date-time"`
+    StartedOn   *time.Time   `gorm:"nullable;default:NULL" json:"startedOn" format:"date-time"`
+    CompletedOn *time.Time   `gorm:"nullable;default:NULL" json:"finishedOn" format:"date-time"`
+    GitBranch   string       `gorm:"size:300;default:'';not null" json:"gitBranch"`
+    Environment null.String  `gorm:"nullable;size:40" json:"environment" swaggertype:"string"`
+    Stage       string       `gorm:"size:40;default:'';not null" json:"stage"`
+    Params      []BuildParam `gorm:"foreignKey:BuildID" json:"params"`
+    IsInvalid   bool         `gorm:"not null;default:false" json:"isInvalid"` 
+    // added 
+    TestResultSummaryCount uint `gorm:"not null" json:"testResultSummaryCount"`
+}
+// new
+type TestResultSummary struct {
+    ArtifactID  uint	  `gorm:"not null;index:testresultsummary_idx_artifact_id" json:"artifactId"`
+    Artifact    *Artifact `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+    BuildID     uint      `gorm:"not null;index:testresultsummary_idx_build_id" json:"buildId"`
+    Build       *Build    `gorm:"foreignKey:BuildID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+    RunCount    uint	  `gorm:"not null" json:"runCount"`
```

Comment:
nit: "Total" would be more suiting IMO. "Run" is somewhat unconventional word

Reply:
I imagined RunCount to be the sum of FailCount and PassCount, leaving out SkipCount.
Then TotalCount would be RunCount + SkipCount.

Would it be better to instead have RunCount be all 3s total, and then rename it TotalCount?

Reply:
Yea. For me at least I understand the "total" nomenclature much better than "run". Also makes more sense that if a value is a sum then it's the sum of all others, not just some others.

We're talking ints here, so like 4 bytes per test result artifact. We can afford it, and even more fields if we so want to.

Reply:
Roger.